File: android/app/src/main/kotlin/run/nuri/getagrip/ui/builder/EveryDaySection.kt
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

package run.nuri.getagrip.ui.builder

import android.Manifest
import android.content.Intent
import android.os.Build
import android.provider.Settings
import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.expandVertically
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.shrinkVertically
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.MaterialTheme
import run.nuri.getagrip.ui.theme.InstrumentSurface as Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.TimePicker
import androidx.compose.material3.rememberTimePickerState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.core.content.ContextCompat
import androidx.core.net.toUri
import run.nuri.getagrip.engine.L10n
import run.nuri.getagrip.engine.ReminderTime
import run.nuri.getagrip.engine.RoutineDraft
import run.nuri.getagrip.ui.components.CapsLabel
import run.nuri.getagrip.ui.components.Chip
import run.nuri.getagrip.ui.components.ChipGrid
import run.nuri.getagrip.ui.components.IntChipRow
import run.nuri.getagrip.ui.l10n.tr
import run.nuri.getagrip.ui.theme.GetAGripTheme
import run.nuri.getagrip.ui.theme.LocalGripPalette
import run.nuri.getagrip.ui.theme.Metrics
import run.nuri.getagrip.ui.theme.Motion
import run.nuri.getagrip.ui.theme.rememberReduceMotion

/// EVERY DAY — how many sessions a day this routine asks for, and when to say something.
///
/// The session count DRIVES the reminder rows beneath it, so "twice a day" is expressed
/// inline. There is no second surface where reminder times live.
@Composable
fun EveryDaySection(
    /// Only what this card draws — see `EveryDayValues`.
    values: EveryDayValues,
    modifier: Modifier = Modifier,
    /// **Whether the permission dialog came back NO**, as a VALUE so this section never touches
    /// a store.
    ///
    /// Not "can we post now": the ask happens on the first SAVE with reminders on, so a first-run
    /// builder is unpermitted simply because it has not asked yet.
    notificationsRefused: Boolean = false,
    update: DraftUpdate,
) {
    val palette = LocalGripPalette.current
    val context = LocalContext.current

    /// From the system: no store owns authorization, and it can change while this screen is open.
    val notificationsBlocked = remember(context) {
        Build.VERSION.SDK_INT >= 33 &&
            ContextCompat.checkSelfPermission(context, Manifest.permission.POST_NOTIFICATIONS) !=
            android.content.pm.PackageManager.PERMISSION_GRANTED
    }

    /// **Blocked is not refused.** The flag records the no; the live check clears the note the
    /// moment permission is granted in system Settings.
    val showDenied = notificationsBlocked && notificationsRefused

    // Repair: a draft whose reminders and session count disagree (older stash, synced merge)
    // would draw fewer rows than promised. Guarded so an untouched document stays clean.
    LaunchedEffect(Unit) {
        update { draft ->
            if (draft.reminders.size != draft.sessionsPerDay) draft.setSessionsPerDay(draft.sessionsPerDay)
            else draft
        }
    }

    Column(modifier.fillMaxWidth(), verticalArrangement = Arrangement.spacedBy(12.dp)) {
        CapsLabel(tr("HOW OFTEN"), Modifier.padding(start = 6.dp))

        Surface(shape = RoundedCornerShape(Metrics.radiusCard), color = palette.card) {
            Column(Modifier.padding(16.dp), verticalArrangement = Arrangement.spacedBy(18.dp)) {
                // Ritual or whenever (Nuri, 2026-08-10). A chip PAIR, not a toggle: two named peers, not one
                // thing switched off.
                ChipGrid(
                    base = 2,
                    modifier = Modifier.semantics { contentDescription = L10n.tr("How often") },
                    content = listOf(
                        { cell: Modifier ->
                            Chip(tr("Daily ritual"), !values.isOnDemand, cell) {
                                update { it.copy(isOnDemand = false) }
                            }
                        },
                        { cell: Modifier ->
                            Chip(tr("Whenever"), values.isOnDemand, cell) {
                                update { it.copy(isOnDemand = true) }
                            }
                        },
                    ),
                )

                if (values.isOnDemand) {
                    // Declined in one sentence. The times stay in the draft, so flipping back restores them.
                    Text(
                        tr("No daily target and no reminders — it waits on Today until you feel like it."),
                        style = MaterialTheme.typography.bodySmall,
                        color = palette.inkTertiary,
                    )
                } else {
                    Column(verticalArrangement = Arrangement.spacedBy(10.dp)) {
                        Text(
                            tr("Sessions a day"),
                            style = MaterialTheme.typography.bodyMedium,
                            fontWeight = FontWeight.SemiBold,
                            color = palette.inkPrimary,
                        )
                        // Via `setSessionsPerDay`, which PARKS removed times: 2 → 1 → 2 restores the user's 19:15,
                        // not the 19:00 default.
                        IntChipRow(
                            values = listOf(1, 2, 3, 4),
                            selection = values.sessionsPerDay,
                            // Bare numeral chips need the container label to make "2" a sentence.
                            modifier = Modifier.semantics { contentDescription = L10n.tr("Sessions a day") },
                        ) { count -> update { it.setSessionsPerDay(count) } }
                    }

                    Column(verticalArrangement = Arrangement.spacedBy(14.dp)) {
                        ToggleRow(
                            title = tr("Remind me"),
                            checked = values.remindersEnabled,
                            explainer = null,
                        ) { enabled -> update { it.copy(remindersEnabled = enabled) } }

                        // The times follow the toggle; the draft keeps them either way. `Motion.state` — see
                        // `TargetBandRow`.
        AnimatedVisibility(
            visible = values.remindersEnabled,
            enter = expandVertically(Motion.state(rememberReduceMotion())) +
                fadeIn(Motion.state(rememberReduceMotion())),
            exit = shrinkVertically(Motion.state(rememberReduceMotion())) +
                fadeOut(Motion.state(rememberReduceMotion())),
        ) {
                            Column(verticalArrangement = Arrangement.spacedBy(14.dp)) {
                                // Bounded by the LIST too: a mismatched draft loses a row, never crashes.
                                val rows = minOf(values.sessionsPerDay, values.reminders.size)
                                repeat(rows) { index ->
                                    ReminderRow(
                                        index = index,
                                        time = values.reminders[index],
                                    ) { newTime ->
                                        // NOT sorted or deduped here: reordering under the finger would swap the
                                        // row being edited. `RoutineDraft.normalized` tidies on save.
                                        update { draft ->
                                            if (index !in draft.reminders.indices) draft
                                            else draft.copy(reminders = draft.reminders.toMutableList()
                                                .also { it[index] = newTime })
                                        }
                                    }
                                }
                                if (showDenied) {
                                    // Denied ends the NOTIFICATION, never the setting: reminders stay ON, so
                                    // relenting in Settings later just works.
                                    DeniedRow()
                                } else if (notificationsBlocked) {
                                    // Not yet asked: saying so makes the OS dialog after Save expected, not an
                                    // ambush — why the ask is contextual.
                                    Text(
                                        tr("Android will ask to allow notifications when you save."),
                                        style = MaterialTheme.typography.bodySmall,
                                        color = palette.inkTertiary,
                                    )
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}

@Composable
private fun ReminderRow(
    index: Int,
    time: ReminderTime,
    onChange: (ReminderTime) -> Unit,
) {
    val palette = LocalGripPalette.current
    var picking by remember { mutableStateOf(false) }
    // Read outside the semantics lambda, which is not composable.
    val spoken = tr("Session %d reminder time", index + 1) + ", " + time.displayText()

    Row(
        Modifier
            .fillMaxWidth()
            .heightIn(min = 44.dp)
            .clickable { picking = true }
            .semantics(mergeDescendants = true) {
                role = Role.Button
                contentDescription = spoken
            },
        horizontalArrangement = Arrangement.spacedBy(12.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        CapsLabel(tr("SESSION %d", index + 1), Modifier.weight(1f))
        Text(
            time.displayText(),
            style = MaterialTheme.typography.titleMedium,
            fontWeight = FontWeight.SemiBold,
            color = palette.inkPrimary,
        )
    }

    if (picking) {
        TimePickerDialog(
            initialHour = time.hour,
            initialMinute = time.minute,
            onDismiss = { picking = false },
        ) { hour, minute ->
            picking = false
            onChange(ReminderTime(hour = hour, minute = minute))
        }
    }
}

/// Material 3 ships a `TimePicker` but no dialog around it, so this is the house wrapper.
@OptIn(androidx.compose.material3.ExperimentalMaterial3Api::class)
@Composable
private fun TimePickerDialog(
    initialHour: Int,
    initialMinute: Int,
    onDismiss: () -> Unit,
    onConfirm: (Int, Int) -> Unit,
) {
    val palette = LocalGripPalette.current
    val state = rememberTimePickerState(
        initialHour = initialHour,
        initialMinute = initialMinute,
        // The system's 12/24-hour setting, as `ReminderTime.displayText` follows.
        is24Hour = android.text.format.DateFormat.is24HourFormat(LocalContext.current),
    )
    AlertDialog(
        onDismissRequest = onDismiss,
        containerColor = palette.card,
        title = { Text(tr("Reminder time"), color = palette.inkPrimary) },
        text = { TimePicker(state = state) },
        confirmButton = {
            TextButton(onClick = { onConfirm(state.hour, state.minute) }) { Text(tr("Set")) }
        },
        dismissButton = { TextButton(onClick = onDismiss) { Text(tr("Cancel")) } },
    )
}

@Composable
private fun DeniedRow() {
    val palette = LocalGripPalette.current
    val context = LocalContext.current
    Column(verticalArrangement = Arrangement.spacedBy(10.dp)) {
        Text(
            tr("Notifications are off for Get a Grip."),
            style = MaterialTheme.typography.bodySmall,
            fontWeight = FontWeight.Medium,
            color = palette.inkSecondary,
        )
        Text(
            tr("Open Settings"),
            style = MaterialTheme.typography.bodyMedium,
            fontWeight = FontWeight.SemiBold,
            color = palette.graphite,
            modifier = Modifier
                .heightIn(min = 44.dp)
                .clickable {
                    context.startActivity(
                        Intent(
                            Settings.ACTION_APPLICATION_DETAILS_SETTINGS,
                            "package:${context.packageName}".toUri(),
                        ),
                    )
                }
                .padding(vertical = 12.dp)
                .semantics { role = Role.Button },
        )
    }
}

@Preview(name = "EveryDaySection", showBackground = true, widthDp = 380)
@Composable
private fun EveryDaySectionPreview() {
    GetAGripTheme {
        var draft by remember { mutableStateOf(RoutineDraft.starter) }
        Column(Modifier.padding(16.dp)) {
            EveryDaySection(EveryDayValues.of(draft)) { draft = it(draft) }
        }
    }
}

File: Sources/UI/Routine/EveryDaySection.swift
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

import SwiftUI
import UIKit
import UserNotifications

/// EVERY DAY — how many sessions a day this routine asks for, and when to remind you.
///
/// The session count DRIVES the reminder rows underneath it, so "twice a day" is set
/// inline in the document, not on a scheduling screen. There is no second surface where
/// reminder times live.
struct EveryDaySection: View, Equatable {
    /// The write path. Everything drawn comes from `schedule` — see `BuilderInputs`.
    let access: DraftAccess
    /// `draft.schedule` (the plan blanked) as a value, so this card compares on the every-day
    /// fields alone and an edit inside a set never re-runs its date pickers.
    let schedule: RoutineDraft

    nonisolated static func == (a: Self, b: Self) -> Bool { a.schedule == b.schedule }

    @Environment(\.scenePhase) private var scenePhase
    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    /// From the notification centre, not a store: nothing there owns authorization, and it
    /// can change WHILE this sheet is open (a trip to iOS Settings and back).
    @State private var authStatus: UNAuthorizationStatus = .notDetermined

    /// A FIXED mid-January day carries the hour into the `DatePicker` — the date
    /// `ReminderTime.displayText()` uses — so DST can never shift the hour.
    private static let referenceDay = DateComponents(year: 2001, month: 1, day: 15)

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            // A PLAIN row, never a `Section` header: plain-style headers pin.
            CapsLabel(String(localized: "HOW OFTEN"))
                .padding(.leading, 6)

            MaterialCard(surface: .flat) {
                VStack(alignment: .leading, spacing: 18) {
                    kindBlock
                    if schedule.isOnDemand {
                        // The times are KEPT in the draft — flipping back to a ritual restores
                        // them — so nothing is destroyed, only quiet.
                        Text("No daily target and no reminders — it waits on Today until you feel like it.")
                            .font(.system(.footnote))
                            .foregroundStyle(Ink.tertiary)
                            .fixedSize(horizontal: false, vertical: true)
                    } else {
                        sessionsBlock
                        reminderBlock
                    }
                }
            }
        }
        .animation(revealAnimation, value: schedule.isOnDemand)
        .onAppear {
            // Repair: a draft whose reminder list and count disagree (an older stash, a
            // merge) would draw fewer rows than promised. The guard keeps an untouched
            // document from turning dirty.
            if schedule.reminders.count != schedule.sessionsPerDay {
                access.mutate { $0.setSessionsPerDay($0.sessionsPerDay) }
            }
        }
        .task { await refreshAuthorization() }
        .onChange(of: scenePhase) { _, phase in
            if phase == .active { Task { await refreshAuthorization() } }
        }
    }

    // MARK: - Kind

    /// Ritual or whenever (Nuri, 2026-08-10). A chip pair, not a toggle: two named peers,
    /// not one thing switched off.
    private var kindBlock: some View {
        VStack(alignment: .leading, spacing: 10) {
            ChipGrid(base: 2) {
                Chip(title: String(localized: "Daily ritual"), isSelected: !schedule.isOnDemand) {
                    access.mutate { $0.isOnDemand = false }
                }
                Chip(title: String(localized: "Whenever"), isSelected: schedule.isOnDemand) {
                    access.mutate { $0.isOnDemand = true }
                }
            }
        }
        .sensoryFeedback(.selection, trigger: schedule.isOnDemand)
        .accessibilityElement(children: .contain)
        .accessibilityLabel(String(localized: "How often"))
    }

    // MARK: - Sessions a day

    private var sessionsBlock: some View {
        VStack(alignment: .leading, spacing: 10) {
            Text("Sessions a day")
                .font(.system(.subheadline, weight: .semibold))
                .foregroundStyle(Ink.primary)

            IntChipRow(values: [1, 2, 3, 4], selection: sessionsBinding)
        }
        // The chips are bare numerals; the container label makes "2" a sentence.
        .accessibilityElement(children: .contain)
        .accessibilityLabel(String(localized: "Sessions a day"))
    }

    /// Never writes `sessionsPerDay` directly: `setSessionsPerDay` PARKS removed times, so
    /// 2 → 1 → 2 restores the user's own 19:15, not the 19:00 default.
    private var sessionsBinding: Binding<Int> {
        Binding(get: { schedule.sessionsPerDay },
                set: { new in access.mutate { $0.setSessionsPerDay(new) } })
    }

    // MARK: - Reminders

    private var reminderBlock: some View {
        VStack(alignment: .leading, spacing: 14) {
            Toggle("Remind me", isOn: access.binding(\.remindersEnabled, current: schedule.remindersEnabled))
                .font(.system(.subheadline, weight: .semibold))
                .foregroundStyle(Ink.primary)
                .tint(Accent.graphite)

            // The times follow the toggle rather than sitting inert; the draft keeps
            // them either way, so turning reminders back on restores the schedule.
            if schedule.remindersEnabled {
                ForEach(0..<rowCount, id: \.self) { index in
                    reminderRow(index)
                }

                if authStatus == .denied {
                    deniedRow
                }
            }
        }
        .animation(revealAnimation, value: schedule.remindersEnabled)
        .animation(revealAnimation, value: rowCount)
    }

    /// Bounded by the list, never the count alone: a mismatched draft loses a row, never
    /// crashes.
    private var rowCount: Int {
        min(schedule.sessionsPerDay, schedule.reminders.count)
    }

    private func reminderRow(_ index: Int) -> some View {
        HStack(spacing: 12) {
            CapsLabel(String(localized: "SESSION \(index + 1)"))
            Spacer(minLength: 8)
            DatePicker("Session \(index + 1) reminder time",
                       selection: timeBinding(index),
                       displayedComponents: .hourAndMinute)
                .datePickerStyle(.compact)
                .labelsHidden()
                .accessibilityLabel(String(localized: "Session \(index + 1) reminder time"))
        }
    }

    /// Denied is a dead end for the notification, not the setting: reminders stay ON in the
    /// draft, so allowing them later in iOS Settings just works.
    private var deniedRow: some View {
        VStack(alignment: .leading, spacing: 10) {
            Text("Notifications are off for Get a Grip.")
                .font(.system(.footnote, weight: .medium))
                .foregroundStyle(Ink.secondary)
                .fixedSize(horizontal: false, vertical: true)

            OpenSettingsButton()
        }
    }

    // MARK: - Bridging ReminderTime ↔ Date

    /// `ReminderTime` stores minutes from midnight; a `DatePicker` wants a `Date`. Only the
    /// hour crosses; display is the picker's own locale formatting.
    private func timeBinding(_ index: Int) -> Binding<Date> {
        Binding(
            get: {
                let reminders = schedule.reminders
                guard index < reminders.count else { return date(for: ReminderTime(hour: 8, minute: 0)) }
                return date(for: reminders[index])
            },
            set: { newDate in
                let parts = Calendar.current.dateComponents([.hour, .minute], from: newDate)
                // NOT sorted or deduped here: reordering under the finger would swap the
                // row being edited. `RoutineDraft.normalized` tidies on the way to the store.
                access.mutate { draft in
                    guard index < draft.reminders.count else { return }
                    draft.reminders[index] = ReminderTime(hour: parts.hour ?? 0, minute: parts.minute ?? 0)
                }
            })
    }

    private func date(for time: ReminderTime, calendar: Calendar = .current) -> Date {
        var comps = Self.referenceDay
        comps.hour = time.hour
        comps.minute = time.minute
        return calendar.date(from: comps) ?? Date(timeIntervalSince1970: 0)
    }

    private var revealAnimation: Animation {
        Motion.state(reduceMotion)
    }

    private func refreshAuthorization() async {
        authStatus = await UNUserNotificationCenter.current().notificationSettings().authorizationStatus
    }
}

/// The one control that needs `openURL`, and so the only view that may STORE it.
///
/// `@Environment(\.openURL)`'s identity moves with the presentation, like `dismiss` (see
/// `BuilderDocument`). Measured 2026-08-18: with a builder field focused, every keystroke
/// logged `EveryDaySection: _openURL changed` three times, rebuilding the chips, reminder
/// rows and `DatePicker`s for a button usually not even on screen. Held by a leaf, the
/// blast radius is the one thing that depends on it.
private struct OpenSettingsButton: View {
    @Environment(\.openURL) private var openURL

    var body: some View {
        SecondaryGlassButton(title: String(localized: "Open Settings")) {
            if let url = URL(string: UIApplication.openSettingsURLString) { openURL(url) }
        }
    }
}
